terraform/transform_reference.go

```diff
@@ -45,6 +43,22 @@ type GraphNodeAttachDependencies interface {
 	AttachDependencies([]addrs.ConfigResource)
 }
 
+// GraphNodeAttachDependsOn records all resources that are transitively
+// referenced through depends_on in the configuration. This is used by data
+// resources to determine if they can be read during the plan, or if they need
+// to be further delayed until apply.
+// We can only use an addrs.ConfigResource address here, because modules are
+// not yet expended in the graph. While this will cause some extra data
+// resources to show in the plan when their depends_on references may be in
+// unrelated module instances, the fact that it only happens when there are any
+// resource updates pending means we can still avoid the problem of the
+// "perpetual diff"
+type GraphNodeAttachDependsOn interface {
+	GraphNodeConfigResource
+	AttachDependsOn([]addrs.ConfigResource)
```

Comment:
Given that "depends on" already has a meaning in our language, might we name this something like `AttachResourceDependencies` instead, to make it a bit clearer that it's getting something _derived from_ `depends_on` rather than getting `depends_on` itself?

Reply:
We already have something like that for _actual_ dependencies though, which is why I ended up on this sort of awkward name. I do want this to be clear that it is derived from `depends_on` in some way, and not the result of normal dependencies with visible changes, but am open to a better name.

Reply:
Since this is specifically about _resource dependencies_ rather than arbitrary dependencies, I'd suggest `AttachResourceDependencies` as a subtle (admittedly) way to differentiate this.